File: library/src/main/java/me/proxer/library/api/forum/TopicEndpoint.java
package me.proxer.library.api.forum;

import lombok.Setter;
import lombok.experimental.Accessors;
import me.proxer.library.api.PagingLimitEndpoint;
import me.proxer.library.api.ProxerCall;
import me.proxer.library.entity.forum.Topic;

import javax.annotation.Nullable;

/**
 * Endpoint for retrieving a forum topic with its posts.
 *
 * @author Ruben Gees
 */
@Accessors(fluent = true)
public final class TopicEndpoint implements PagingLimitEndpoint<Topic> {

    private final InternalApi internalApi;

    private final String id;

    /**
     * {@inheritDoc}
     */
    @Nullable
    @Setter(onMethod = @__({@Override}))
    private Integer page;

    /**
     * {@inheritDoc}
     */
    @Nullable
    @Setter(onMethod = @__({@Override}))
    private Integer limit;

    TopicEndpoint(final InternalApi internalApi, final String id) {
        this.internalApi = internalApi;
        this.id = id;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public ProxerCall<Topic> build() {
        return internalApi.topic(id, page, limit);
    }
}

File: library/src/main/kotlin/me/proxer/library/api/forum/TopicEndpoint.kt
package me.proxer.library.api.forum

import me.proxer.library.api.PagingLimitEndpoint
import me.proxer.library.api.ProxerCall
import me.proxer.library.entity.forum.Topic

/**
 * Endpoint for retrieving a forum topic with its posts.
 *
 * @author Ruben Gees
 */
class TopicEndpoint internal constructor(
    private val internalApi: InternalApi,
    private val id: String
) : PagingLimitEndpoint<Topic> {

    private var page: Int? = null
    private var limit: Int? = null

    override fun page(page: Int?) = this.apply { this.page = page }
    override fun limit(limit: Int?) = this.apply { this.limit = limit }

    override fun build(): ProxerCall<Topic> {
        return internalApi.topic(id, page, limit)
    }
}
